src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.OptionTracking.cs

```diff
@@ -0,0 +1,199 @@
+                    if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"GetSockOpt({level},{name}) returned errorCode:{errorCode}");
+                }
+            }
+        }
+
+        internal void SetTrackedSocketOptions(ReadOnlySpan<int> values, LingerOption? lingerOption)
+        {
+            Debug.Assert(values.Length == TrackableOptionCount);
+            int lingerMask = GetMask(TrackableSocketOptions.SO_LINGER);
+            if (lingerOption is not null)
+            {
+                Debug.Assert((_trackedOptions & lingerMask) == lingerMask);
+                SocketError errorCode = SocketPal.SetLingerOption(this, lingerOption);
+                if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"SetLingerOption returned errorCode:{errorCode}");
+            }
+
+            int trackedOptions = _trackedOptions & ~lingerMask;
+
+            for (int i = 0; i < values.Length; i++)
+            {
+                int mask = 1 << i;
+                if ((trackedOptions & mask) == mask)
+                {
+                    TrackableSocketOptions tracked = (TrackableSocketOptions)(i + 1);
+                    (SocketOptionName name, SocketOptionLevel level) = ToSocketOptions(tracked);
+                    SocketError errorCode = SocketPal.SetSockOpt(this, level, name, values[i]);
+                    if (NetEventSource.Log.IsEnabled() && errorCode != SocketError.Success) NetEventSource.Info(this, $"GetSockOpt({level},{name}) returned errorCode:{errorCode}");
+                }
+            }
+        }
+
+        private static int GetMask(TrackableSocketOptions tracked) => 1 << ((int)tracked - 1);
+
+        // Relevant option names and values taken from Windows headers.
+        private enum TrackableSocketOptions
+        {
+            None = 0,
+            IP_TOS,
+            IP_TTL,
+            IPV6_PROTECTION_LEVEL,
+            IPV6_V6ONLY,
+            TCP_NODELAY,
+            TCP_EXPEDITED_1122,
+            TCP_KEEPALIVE,
+            TCP_FASTOPEN,
+            TCP_KEEPCNT,
+            TCP_KEEPINTVL,
+            SO_DEBUG,
+            SO_ACCEPTCONN,
+            SO_REUSEADDR,
+            SO_KEEPALIVE,
+            SO_DONTROUTE,
+            SO_USELOOPBACK,
+            SO_LINGER,
+            SO_OOBINLINE,
+            SO_DONTLINGER,
+            SO_EXCLUSIVEADDRUSE,
+            SO_SNDBUF,
+            SO_RCVBUF,
+            SO_SNDLOWAT,
+            SO_RCVLOWAT,
+            SO_SNDTIMEO,
+            SO_RCVTIMEO
+        }
+
+        internal static int TrackableOptionCount => (int)TrackableSocketOptions.SO_RCVTIMEO;
+
+        private static TrackableSocketOptions ToTrackableSocketOptions(SocketOptionName name, SocketOptionLevel level)
+            => ((int)name, level) switch
+            {
+                (3, SocketOptionLevel.IP) => TrackableSocketOptions.IP_TOS,
```

Comment:
Nit: Can use use the enum names instead of values here? (same for `ToSocketOptions` below)

Reply:
The problem is that some options have multiple names where the "best" name would depend on the SocketOptionLevel. This would need a manual edit of the generated code since it only cares about the values as they are defined in the Windows headers.